src/transports/http.c

```diff
@@ -1017,27 +1017,21 @@ static int http_stream_read(
 	http_subtransport *t = OWNING_SUBTRANSPORT(s);
 	parser_context ctx;
 	size_t bytes_parsed;
+	git_buf request = GIT_BUF_INIT;
+	int error = 0;
 
 replay:
 	*bytes_read = 0;
 
 	assert(t->connected);
 
 	if (!s->sent_request) {
-		git_buf request = GIT_BUF_INIT;
-
+		git_buf_clear(&request);
 		clear_parser_state(t);
 
-		if (gen_request(&request, s, 0) < 0)
-			return -1;
-
-		if (git_stream__write_full(t->server.stream, request.ptr,
-					   request.size, 0) < 0) {
-			git_buf_dispose(&request);
-			return -1;
-		}
-
-		git_buf_dispose(&request);
```

Comment:
Sorry, I could have been more clear.  We're not explicitly calling `realloc`, but we are doing that implicitly, since we were defining the `request` buffer inside the `replay` block.  So we would create a new buffer for every replay, which is not necessary.  Instead we can clear the existing buffer and just put the new URI inside of it.